From 462c779e446a76814869afef5d15c56a076bdacf Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Wed, 06 Nov 2019 12:51:24 +0200
Subject: [PATCH] [SW-0] hl: precompile linker script

This way we can use toolchain macros to behave properly

Change-Id: Id14fa26a4dbbbd0681fdc065a021c53e6b6679d9
---

diff --git a/Makefile b/Makefile
index 00cbbcd..2da177a 100644
--- a/Makefile
+++ b/Makefile
@@ -15,7 +15,8 @@
 
 SFLAGS := $(CFLAGS)
 
-LDFLAGS := -g -T $(HL_DEV)/preboot.lds -nostdlib -static
+LD_SCRIPT := $(HL_DEV)/preboot.lds
+LDFLAGS := -g -T $(LD_SCRIPT) -nostdlib -static
 LDFLAGS += -L ../../host/opt/ext-toolchain/lib/gcc/aarch64-linux-gnu/6.4.1
 LDLIBS := -lgcc
 
@@ -35,8 +36,8 @@
 preboot.bin: preboot
 	$(OBJCOPY) -O binary -R .note.gnu.build-id $< $@
 
-preboot: $(PREBOOT_OBJECTS)
-	$(CC) $(LDFLAGS) -o $@ -Wl,--start-group $^ $(LDLIBS) -Wl,--end-group
+preboot: $(PREBOOT_OBJECTS) $(LD_SCRIPT)
+	@$(CC) $(LDFLAGS) -o $@ -Wl,--start-group $(PREBOOT_OBJECTS) $(LDLIBS) -Wl,--end-group
 
 $(AUTO_GENERATED_FILE): localversion-habanalabs $(PREBOOT_SRCS_C) $(PREBOOT_SRCS_S)
 	@( printf '#define PREBOOT_VER "%s"\n' '$(shell cut -d- -f2- localversion-habanalabs)' ) > $@
@@ -49,6 +50,9 @@
 common/pcie_phy_fw.i: common/pcie_phy_fw.hex
 	@gcc -x c -E -P -nostdinc -o $@ $<
 
+$(LD_SCRIPT): $(LD_SCRIPT).S
+	@$(CC) -E -P -nostdinc -o $@ $<
+
 %.o: %.c $(AUTO_GENERATED_FILE)
 	$(CC) $(CFLAGS) -o $@ $<
 
diff --git a/gaudi/preboot.lds b/gaudi/preboot.lds
deleted file mode 100644
index 01774e2..0000000
--- a/gaudi/preboot.lds
+++ /dev/null
@@ -1,23 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-MEMORY { .sram : ORIGIN = 0x7FFBFE0000,
-		LENGTH = 0x10000 }
-
-OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
-OUTPUT_ARCH(aarch64)
-ENTRY(_start)
-SECTIONS
-{
-	.text : {
-		. = ALIGN(8);
-		common/start.o (.text*)
-		* (.text*)
-	} >.sram
-
-	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
-}
diff --git a/gaudi/preboot.lds.S b/gaudi/preboot.lds.S
new file mode 100644
index 0000000..ea6a573
--- /dev/null
+++ b/gaudi/preboot.lds.S
@@ -0,0 +1,22 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+MEMORY { .sram : ORIGIN = 0x7FFBFE0000, LENGTH = 0x10000 }
+
+OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
+OUTPUT_ARCH(aarch64)
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(8);
+		common/start.o (.text*)
+		* (.text*)
+	} >.sram
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
diff --git a/goya/preboot.lds b/goya/preboot.lds
deleted file mode 100644
index ea1f531..0000000
--- a/goya/preboot.lds
+++ /dev/null
@@ -1,23 +0,0 @@
-/*
- * (C) Copyright 2017 HabanaLabs, Ltd.
- * Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-MEMORY { .sram : ORIGIN = 0x7ff0000000,
-		LENGTH = 0x10000 }
-
-OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
-OUTPUT_ARCH(aarch64)
-ENTRY(_start)
-SECTIONS
-{
-	.text : {
-		. = ALIGN(8);
-		common/start.o (.text*)
-		* (.text*)
-	} >.sram
-
-	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
-}
diff --git a/goya/preboot.lds.S b/goya/preboot.lds.S
new file mode 100644
index 0000000..98653bc
--- /dev/null
+++ b/goya/preboot.lds.S
@@ -0,0 +1,22 @@
+/*
+ * (C) Copyright 2017 HabanaLabs, Ltd.
+ * Noam Camus <noamikong@gmail.com>
+ *
+ * SPDX-License-Identifier:     GPL-2.0+
+ */
+
+MEMORY { .sram : ORIGIN = 0x7ff0000000, LENGTH = 0x10000 }
+
+OUTPUT_FORMAT("elf64-littleaarch64", "elf64-littleaarch64", "elf64-littleaarch64")
+OUTPUT_ARCH(aarch64)
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(8);
+		common/start.o (.text*)
+		* (.text*)
+	} >.sram
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
